src/test/java/com/android/volley/toolbox/DiskBasedAsyncCacheTest.java

```diff
@@ -0,0 +1,449 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(sdk = 26)
+public class DiskBasedAsyncCacheTest {
+
+    private static final int MAX_SIZE = 1024 * 1024;
+
+    private DiskBasedAsyncCache cache;
+
+    private AsyncCache.OnCompleteCallback emptyCallback;
+
+    @Rule public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @Rule public ExpectedException exception = ExpectedException.none();
+
+    @Before
+    public void setup() throws IOException {
+        emptyCallback =
+                new AsyncCache.OnCompleteCallback() {
+                    @Override
+                    public void onComplete() {
+                        // do nothing
+                    }
+                };
+        // Initialize empty cache
+        cache = new DiskBasedAsyncCache(temporaryFolder.getRoot(), MAX_SIZE);
+        cache.initialize(emptyCallback);
+    }
+
+    @After
+    public void teardown() {
+        cache = null;
+    }
+
+    @Test
+    public void testEmptyInitialize() throws ExecutionException, InterruptedException {
+        assertNull(getEntry("key").get());
+    }
+
+    @Test
+    public void testPutGetZeroBytes() throws ExecutionException, InterruptedException {
+        final Cache.Entry entry = new Cache.Entry();
+
+        entry.data = new byte[0];
+        entry.serverDate = 1234567L;
+        entry.lastModified = 13572468L;
+        entry.ttl = 9876543L;
+        entry.softTtl = 8765432L;
+        entry.etag = "etag";
+        entry.responseHeaders = new HashMap<>();
+        entry.responseHeaders.put("fruit", "banana");
+        entry.responseHeaders.put("color", "yellow");
+
+        putEntry("my-magical-key", entry).get();
+
+        assertThatEntriesAreEqual(getEntry("my-magical-key").get(), entry);
+        assertThat(getEntry("unknown-key").get(), is(nullValue()));
+    }
+
+    @Test
+    public void testTooLargeEntry() throws ExecutionException, InterruptedException {
+        Cache.Entry entry = randomData(MAX_SIZE - getEntrySizeOnDisk("oversize") + 1);
+        putEntry("oversize", entry).get();
+
+        assertThat(getEntry("oversize").get(), is(nullValue()));
+    }
+
+    @Test
+    public void testMaxSizeEntry() throws ExecutionException, InterruptedException {
+        Cache.Entry entry = randomData(MAX_SIZE - getEntrySizeOnDisk("maxsize") - 1);
+        putEntry("maxsize", entry).get();
+
+        assertThatEntriesAreEqual(getEntry("maxsize").get(), entry);
+    }
+
+    @Test
+    public void testTrimAtThreshold() throws ExecutionException, InterruptedException {
+        // Start with the largest possible entry.
+        Cache.Entry entry = randomData(MAX_SIZE - getEntrySizeOnDisk("maxsize") - 1);
```

Comment:
I think largest size is now MAX_SIZE if `=` was removed from check in putEntry?

Reply:
yep, I'll make that change here and in DiskBasedCacheTest